paddle/fluid/operators/fc_mkldnn_op.cc

```diff
@@ -126,184 +44,80 @@ class FCMKLDNNOpKernel : public paddle::framework::OpKernel<T> {
-    auto weights_memory = mem.weights(w_data);
-    auto bias_memory = mem.bias();
-
-    const std::string key = ctx.op().Input("Out");
-    const std::string key_fc_pd = key + "@fc_pd";
-
-    auto pd =
-        std::static_pointer_cast<mkldnn::inner_product_forward::primitive_desc>(
-            dev_ctx.GetBlob(key_fc_pd));
-
-    PADDLE_ENFORCE(pd != nullptr, "Fail to find key_fc_pd in device context");
-
-    if (w_grad) {
-      auto weights_grad_memory = mem.weights(w_grad_data);
-
-      mkldnn::inner_product_backward_weights::primitive_desc bwd_weight_pd =
-          FcBwdWeightsPrimitiveDesc(md.src(), md.weights(), md.dst(), md.bias(),
-                                    with_bias, *pd, mkldnn_engine);
-
-      auto bwd_weights_prim = mkldnn::inner_product_backward_weights(
-          bwd_weight_pd, src_memory, dst_memory, weights_grad_memory,
-          bias_memory);
-
-      std::vector<mkldnn::primitive> pipeline{bwd_weights_prim};
-      mkldnn::stream(mkldnn::stream::kind::eager).submit(pipeline).wait();
+    std::vector<int> fc_src_tz =
+        paddle::framework::vectorize2int(input->dims());
+    std::vector<int> fc_weights_tz =
+        paddle::framework::vectorize2int(w->dims());
+    std::vector<int> fc_dst_tz =
+        paddle::framework::vectorize2int(output->dims());
+
+    // MKLDNN requires weights layout to be column major.
+    // The values have already been transposed, but the shape needs to be fixed.
+    // It cannot be done during an earlier stage since InferShape verifies
+    // dimensions assuming the weights weren't transposed.
+    std::swap(fc_weights_tz[0], fc_weights_tz[1]);
+
+    auto fc_src_md = platform::MKLDNNMemDesc(
+        fc_src_tz, platform::MKLDNNGetDataType<T>(), input->format());
+    auto fc_src_memory_pd = memory::primitive_desc(fc_src_md, mkldnn_engine);
+    auto fc_src_memory =
+        memory(fc_src_memory_pd, to_void_cast<T>(input->data<T>()));
+
+    auto fc_weights_md = platform::MKLDNNMemDesc(
+        fc_weights_tz, platform::MKLDNNGetDataType<T>(), w->format());
+    auto fc_weights_memory_pd =
+        memory::primitive_desc(fc_weights_md, mkldnn_engine);
+    auto fc_weights_memory =
+        memory(fc_weights_memory_pd, to_void_cast<T>(w->data<T>()));
+
+    auto fc_dst_md = platform::MKLDNNMemDesc(fc_dst_tz, mkldnn::memory::f32,
+                                             mkldnn::memory::format::any);
+
+    std::shared_ptr<memory> fc_bias_memory_p;
+    std::shared_ptr<inner_product_forward::desc> fc_desc_p;
+    if (bias) {
+      std::vector<int> fc_bias_tz =
+          paddle::framework::vectorize2int(bias->dims());
+      auto fc_bias_md = platform::MKLDNNMemDesc(
+          fc_bias_tz, platform::MKLDNNGetDataType<T>(), bias->format());
+      auto fc_bias_memory_pd =
+          memory::primitive_desc(fc_bias_md, mkldnn_engine);
+      fc_bias_memory_p.reset(
+          new memory(fc_bias_memory_pd, to_void_cast<T>(bias->data<T>())));
+
```

Comment:
Maybe for a future reference: in several places, creating MKLDNN memory primitive looks pretty much the same for different data. Do you think you could move it to a function and use it?

Reply:
This is certainly a good idea and can be implemented if we decide to introduce the primitive reuse for FC.